arrow-core-data/src/test/kotlin/arrow/core/CanTest.kt

```diff
@@ -0,0 +1,415 @@
+import arrow.core.test.generators.genK
+import arrow.core.test.generators.genK2
+import arrow.core.test.laws.BicrosswalkLaws
+import arrow.core.test.laws.BifunctorLaws
+import arrow.core.test.laws.BitraverseLaws
+import arrow.core.test.laws.EqK2Laws
+import arrow.core.test.laws.HashLaws
+import arrow.core.test.laws.MonadLaws
+import arrow.core.test.laws.ShowLaws
+import arrow.core.test.laws.TraverseLaws
+import arrow.typeclasses.Eq
+import arrow.typeclasses.Monad
+import io.kotlintest.data.forall
+import io.kotlintest.properties.Gen
+import io.kotlintest.properties.forAll
+import io.kotlintest.shouldBe
+
+class CanTest : UnitSpec() {
+
+  init {
+
+    val intCanMonad: Monad<CanPartialOf<Int>> = Can.monad(Int.semigroup())
+
+    val EQ = Can.eq(Eq.any(), Eq.any())
+
+    testLaws(
+      EqK2Laws.laws(Can.eqK2(), Can.genK2()),
+      BifunctorLaws.laws(Can.bifunctor(), Can.genK2(), Can.eqK2()),
+      ShowLaws.laws(Can.show(String.show(), Int.show()), EQ, Gen.can(Gen.string(), Gen.int())),
+      MonadLaws.laws(
+        Can.monad(Int.semigroup()),
+        Can.functor(),
+        Can.applicative(Int.semigroup()),
+        Can.monad(Int.semigroup()),
+        Can.genK(Gen.int()),
+        Can.eqK(Int.eq())
+      ),
+      TraverseLaws.laws(Can.traverse(),
+        Can.applicative(Int.semigroup()),
+        Can.genK(Gen.int()),
+        Can.eqK(Int.eq())
+      ),
+      HashLaws.laws(Can.hash(String.hash(), Int.hash()), Gen.can(Gen.string(), Gen.int()), Can.eq(String.eq(), Int.eq())),
+      BitraverseLaws.laws(Can.bitraverse(), Can.genK2(), Can.eqK2()),
+      BicrosswalkLaws.laws(Can.bicrosswalk(), Can.genK2(), Can.eqK2())
+    )
+
+    "performant type checkers should match each type" {
+      fun Can<Any, Any>.expectToBe(neither: Boolean = false, left: Boolean = false, right: Boolean = false, both: Boolean = false) {
+        isNeither shouldBe neither
+        isRight shouldBe right
+        isLeft shouldBe left
+        isBoth shouldBe both
+      }
+      Can.Neither.expectToBe(neither = true)
+      Can.Right(12).expectToBe(right = true)
+      Can.Left(11).expectToBe(left = true)
+      Can.Both(11, 12).expectToBe(both = true)
+    }
+
+    "factory functions create the expected types" {
+      Can.neither() shouldBe Can.Neither
+      allEqual<Int, String>(
+        { Can.left(a) to Can.Left(a) },
+        { Can.right(b) to Can.Right(b) },
+        { Can.Both(a, b) to Can.Both(a, b) },
+        { a.toLeftCan() to Can.Left(a) },
+        { b.toRightCan() to Can.Right(b) },
+        { (a to b).toCan() to Can.Both(a, b) },
+        { (a toT b).toCan() to Can.Both(a, b) },
+        { Ior.Left(a).toCan() to Can.Left(a) },
+        { Ior.Right(b).toCan() to Can.Right(b) },
+        { Ior.Both(a, b).toCan() to Can.Both(a, b) },
+        { None.toRightCan() to Can.Neither },
+        { Some(b).toRightCan() to Can.Right(b) },
+        { None.toLeftCan() to Can.Neither },
+        { Some(a).toLeftCan() to Can.Left(a) }
+      )
```

Comment:
I'm not 100% sold on how I've implemented multi-case tests on this suite as they are hard to pin point failures when they happen since the whole groups fails. The idea of `allEqual` was to get better info, but doesn't work. 
I'm open to suggestions, although I may look at them on a follow-up PR.